tests/test_zero_data_parallel/test_sharded_optim_v2.py

```diff
@@ -34,47 +35,77 @@ def run_step(model, optimizer, data, label, criterion, enable_autocast=False):
     optimizer.step()
 
 
-def run_dist(rank, world_size, port, cpu_offload, shard_strategy):
+def _run_dist(rank, world_size, port, cpu_offload, shard_strategy, use_cpuadam):
     colossalai.launch(config=CONFIG, rank=rank, world_size=world_size, host='localhost', port=port, backend='nccl')
     test_models = ['repeated_computed_layers', 'resnet18', 'bert']
     shard_strategy = shard_strategy()
+
+    if use_cpuadam and cpu_offload is False:
+        return
+
     for model_name in test_models:
         get_components_func = non_distributed_component_funcs.get_callable(model_name)
-        model, train_dataloader, test_dataloader, optimizer_class, criterion = get_components_func()
+        model, train_dataloader, _, optimizer_class, criterion = get_components_func()
         model = model(checkpoint=True).cuda()
         zero_model = ShardedModelV2(copy.deepcopy(model),
                                     shard_strategy,
                                     offload_config=dict(device='cpu') if cpu_offload else None)
         if dist.get_world_size() > 1:
             model = DDP(model)
         lr = 1e-3
-        optim = optimizer_class(model.parameters(), lr=lr)
-        sharded_optim = ShardedOptimizerV2(zero_model,
-                                           optimizer_class,
-                                           cpu_offload=cpu_offload,
-                                           initial_scale=2**5,
-                                           lr=lr)
+        if use_cpuadam:
+            optim = torch.optim.Adam(model.parameters(), lr=lr)
+            sharded_optim = ShardedOptimizerV2(zero_model, CPUAdam, cpu_offload=cpu_offload, initial_scale=2**5, lr=lr)
+        else:
+            optim = optimizer_class(model.parameters(), lr=lr)
+            sharded_optim = ShardedOptimizerV2(zero_model,
+                                               optimizer_class,
+                                               cpu_offload=cpu_offload,
+                                               initial_scale=2**5,
+                                               lr=lr)
         for i, (data, label) in enumerate(train_dataloader):
-            if i > 2:
+            #FIXME() if i > 5, the unittest will fail
+            if i > 3:
```

Comment:
@ver217 @FrankLeeeee after 5 iterations, the diff increases.

Reply:
AssertionError: tensor([-0.0243,  0.0190, -0.0040,  ..., -0.0115,  0.0364, -0.0387],
E              device='cuda:1', grad_fn=<SplitBackward0>) vs tensor([-0.0225,  0.0222, -0.0005,  ..., -0.0128,  0.0355, -0.0404],
E              device='cuda:1')

Reply:
I think we may use fp16 ddp model here. If you print `loss`, you may find that their losses are also different.

Reply:
Can we accept the diff?

Reply:
I think can compare with torch amp instead of pure fp16 or pure fp32.

Reply:
> I think can compare with torch amp instead of pure fp16 or pure fp32.

Yes. But torch amp is `O1` and our `ShardedOptimizer` is `O2`. I think the diff may still exist.

Reply:
Perhaps, can check for torhc amp and apex O2.

Reply:
I merge the MR, you guys please fix the precision issues with other MRs.